urr.c

```diff
@@ -75,6 +75,52 @@ void urr_update(struct urr *urr, struct gtp5g_dev *gtp)
     }
 }
 
+void urr_quota_exhaust_action(struct urr *urr, struct gtp5g_dev *gtp)
+{
+    struct hlist_head *head;
+    struct pdr *pdr;
+    char seid_urr_id_hexstr[SEID_U32ID_HEX_STR_LEN] = {0};
+    int i = 0;
+    // urr stop measurement
+    urr->info |= URR_INFO_INAM;
+    urr->quota_exhausted = true;
+    urr->pdrids = kzalloc(0xff * sizeof(u16), GFP_KERNEL);
+    urr->actions = kzalloc(0xff * sizeof(u8), GFP_KERNEL);
+
+    seid_urr_id_to_hex_str(urr->seid, urr->id, seid_urr_id_hexstr);
+    head = &gtp->related_urr_hash[str_hashfn(seid_urr_id_hexstr) % gtp->hash_size];
+    //each pdr that associate with the urr drop pkt
+    hlist_for_each_entry_rcu(pdr, head, hlist_related_urr) {
+        if (find_urr_id_in_pdr(pdr, urr->id)) {
+            urr->pdrids[i] = pdr->id;
+            urr->actions[i++] = pdr->far->action;
+
+            pdr->far->action = FAR_ACTION_DROP;
+        }
+    }
+}
+void urr_reverse_quota_exhaust_action(struct urr *urr, struct gtp5g_dev *gtp)
+{
+    struct hlist_head *head;
+    struct pdr *pdr;
+    char seid_urr_id_hexstr[SEID_U32ID_HEX_STR_LEN] = {0};
+    int i;
+    urr->quota_exhausted = false;
+    seid_urr_id_to_hex_str(urr->seid, urr->id, seid_urr_id_hexstr);
+    head = &gtp->related_urr_hash[str_hashfn(seid_urr_id_hexstr) % gtp->hash_size];
+
+    //each pdr that associate with the urr resume it's normal action
+    hlist_for_each_entry_rcu(pdr, head, hlist_related_urr) {
+        if (find_urr_id_in_pdr(pdr, urr->id)) {
+            for (i = 0; i < sizeof(urr->pdrids);i++)
```

Comment:
the type urr->pdrids is (u16 *), sizeof(urr->pdrids) will be 4 or 8 depending on machine. Should record the real length in urr.